index.bs

```diff
@@ -304,9 +304,15 @@ partial interface NeuralNetworkContext {
 ## Model ## {#api-model}
 <script type=idl>
 enum CompilationPreference {
+  // No preference. Let the user agent decide the type of compilation most 
+  // suitable and broadly available. This is the default option.
+  "default",
+  // Indicate a preference for a type of compilation that prioritizes power 
+  // consumption over performance.
   "low-power",
-  "fast-answer",
-  "sustained-speed"
+  // Indicate a preference for a type of compilation that prioritizes high 
+  // performance over power consumption.
+  "high-performance"
```

Comment:
"fast-answer" prioritizes maximizing the responsiveness of a single answer. "sustained-speed" prioritizes maximizing the throughput of successive frames.

I propose to create a separate issue to discuss about substituting them with "high-performance".

Reply:
Thanks for your clarification. I had some difficulties distinguishing them just by interpreting the words alone. If the intention is to distinguish between the user's preferences between performance and throughput, which dose make sense, my advice is to spell it out directly e.g. 'high-performance' and 'high-throughput' to avoid having interpret the more abstract words. Could that work? 

Could you also give a few sample scenarios where 'high-throughput' is preferred over 'high-performance'? Concrete examples will be very handy for the description text.

Reply:
> I had some difficulties distinguishing them just by interpreting the words alone.

Apologize for the confusing.

> If the intention is to distinguish the user's preferences between performance and throughput, which does make sense

The intention is to distinguish between latency and throughput. Would it be clearer by using "low-latency" and "high-throughput"?

Reply:
'low-latency' sounds fine too.

Reply:
The compiler may need to a number of different tradeoffs. "Latency" vs. "throughput" is one tradeoff. "Power consumption" vs "performance" is a different tradeoff. What are the ones we need to support? If we want more than one, then preference should not be a single enum: it should perhaps be a list of options?

Reply:
> If we want more than one, then preference should not be a single enum: it should perhaps be a list of options?

I agree with you. We can solve it by introducing `CompilationOptions` dictionary that could contain multiple preferences. For example
```webidl
interface Model {
  Promise<Compilation> createCompilation(optional CompilationOptions options = {});
};

dictionary CompilationOptions {
    PowerPreference powerPreference;
    PerformancePreference performancePreference;
};

enum PowerPreference {
  "low-power",
  "high-performance"
};

enum PerformancePreference {
  "low-latency",
  "high-throughput"
};
```

@gramalingam, what do you think?

Reply:
@huningxin In general, I think we should think through how a browser or the OS may implement these options. It may not be very useful to define options that cannot be achieved in reality. Take for example, I can see how the power preferences can be supported in the OS today as there is already a way to select a low-power adapter if the hardware supports multiple adapters. I can't quite yet imagine how to accomplish the distinction between speed and throughput. This is much harder in reality as it requires special treatments of graph execution process e.g. batching etc, to maximize the graph compute throughput and avoid stutters.

How about we stick with just PowerPreference for now? This will also be consistent with WebGL _WebGLPowerPreference_ options.

Reply:
I tend to agree. Quoting Joshua Bloch: When in doubt, leave it out.

If there is a fundamental theorem of API design, this is it. It applies equally to functionality, classes, methods, and parameters. Every facet of an API should be as small as possible, but no smaller. You can always add things later, but you can't take them away. Minimizing conceptual weight is more important than class- or method-count.

Reply:
@anssiko I was just discussing the WebGL _failIfMajorPerformanceCaveat_ flag with @rafaelcintron and how it could apply to our ML audience. We think we should leave that flag out for now too for simplicity. 

Unlike in typical graphics scenario with WebGL where a GPU is almost always a clear winner over a CPU when it comes to rendering perf of heavy graphics, ML perceived performance could be model-dependent. GPU is of clear advantage with deep networks that require significant compute power while there are also many simpler regressions or classifiers that using GPU may never be significantly faster than a CPU, or sometimes even slower.

Reply:
> I can't quite yet imagine how to accomplish the distinction between speed and throughput.

It is actually about how fast the app can get the first response back. For example, when I use WinMLRunner to infer a model, when I choose run it on CPU device, the first iteration time is much faster than GPU device. However, when I run the model repeatedly, the GPU device outperforms CPU device in average performance excluding the first one.  So one  implementation may select CPU device for `fast-answer` and select GPU device for `sustained-speed`. Does that make sense?

I am pasting the WinMLRunner logs of SqueezeNet inference on my machine for your reference. According the results, the first iteration time on CPU is about 31ms, on GPU is about 60ms. However, the average performance on CPU is about 6.9ms, on GPU is 4.37ms.

CPU:
```
First Iteration Performance (load, bind, session creation, and evaluate):
  Load: 10.9661 ms
  Bind: 0.2029 ms
  Session Creation: 11.2625 ms
  Evaluate: 8.8001 ms

Average Performance excluding first iteration. Iterations 2 to 20. (Iterations greater than 1 only bind and evaluate)
  Average Bind: 0.126079 ms
  Average Evaluate: 6.83456 ms
```
GPU:
```
First Iteration Performance (load, bind, session creation, and evaluate):
  Load: 10.1944 ms
  Bind: 1.3036 ms
  Session Creation: 42.8985 ms
  Evaluate: 5.6366 ms

Average Performance excluding first iteration. Iterations 2 to 20. (Iterations greater than 1 only bind and evaluate)
  Average Bind: 0.296279 ms
  Average Evaluate: 4.0754 ms
```

Reply:
@huningxin Thanks for the example. I think I might have misunderstood the intent behind 'fast-answer' and 'sustained-speed' yet again. In our past conversation, I was led to believe that the distinction between the two is simply latency vs. throughput. What you are now describing is latency vs. first-inference-latency, which is a rather different metrics.

GPU is known to have high initial setup cost from necessary resource creation processes e.g. device and resource creations, pipeline state and shader compilation etc. But once the pipeline state is set up, subsequent executions are extremely fast. A customary practice for GPU utilization is to avoid repeating this setup cost by keeping the pipeline state around and not tearing it down after every single inference call. Historically, the browser is known to do a pretty good job at managing this cost while leveraging GPU for both Canvas2d and WebGL context. But the first-iteration latency itself is often mitigated at the app level by simply discarding the first run.

Anyhow, I hope to illustrate a point that performance-based preference is a broad topic and often scenario-based. Within this thread we have now identified at least a handful of them and I'm sure there may be more.

Reply:
@huningxin @RafaelCintron @gramalingam I've pushed out a new commit based on the feedbacks from this thread so far. Please take a look.